src/engine/IdTable.h

```diff
@@ -109,8 +109,14 @@ class IdTableImpl {
       return *reinterpret_cast<const value_type*>(_data + (_row * COLS));
     }
 
-    Id& operator[](size_t i) { return *(_data + _row * COLS + i); }
-    const Id& operator[](size_t i) const { return *(_data + _row * COLS + i); }
+    // access the element that is i steps ahead
+    // used by the parallel sorting in GCC
+    reference operator[](size_t i) {
+      return *reinterpret_cast<value_type*>(_data + (_row + i) * COLS);
```

Comment:
Your math looks wrong to me (and the old math looks right). The table is stored in Row-Major order.
So let's assume a 2×3 table `T = [[0, 1, 2], [3, 4, 5]]` it is stored as `Tflat= [0, 1, 2, 3, 4, 5]`. Now with the original formula `T[1,1] == Tflat[1*3+1] == 4`. With your formula however we would search for it at `Tflat[(1+1)*3]` which is even out of bounds.

Reply:
@niklas88 @floriankramer You are both mixing up `operator[]` for `Row` and `operator[]` for `IdTable::Iterator`. (Florian implemented them in exactly the same way).

`Row::operator[](k)` takes a row and gives us the Id at the k-th column of that very row. This operator is called in Niklas'  `T[1][1]` example since `T[1]` gives us a `Row`.

`IdTable::Iterator::operator[](k)` is supposed to give us ` *(i + k)` where `i` is the current iterator. This means it returns the reference to a row that is `k`  steps away from the row that `i` is currently pointing to (see [cppreference](https://en.cppreference.com/w/cpp/named_req/RandomAccessIterator) ). I think for this case my formula is correct.

I will look into making the iterator a fully specified random access iterator according to this specification tomorrow (e.g. the argument here should be something signed etc). And to be fair I also had to google what this operator should do. But now I am fairly ceratin.

To Niklas' example:
Your out of bounds calculation would apply, if we had an iterator pointing to the element at idx 1. And then accessing 1 step further is indeed out of bounds:
```
auto it = T.begin() + 1 // iterator to second element
const auto& row = it[1] // reference to third element which is indeed out of bounds.
```

Reply:
Ok I think I get it now, didn't really consider that the iterators themselves have an `operator[]` that for `IdTable::iterator` needs to go over rows.